Groundedness improvments and dotnet examples

lemillermicrosoft opened the pull request:
### Motivation and Context
PR Feedback for https://github.com/microsoft/semantic-kernel/pull/1064

### Contribution Checklist
<!-- Before submitting this PR, please make sure: -->
- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

riedgar-ms committed:
Putting in an entity extraction skill

riedgar-ms committed:
More for grounding

riedgar-ms committed:
Fiddling

riedgar-ms committed:
Putting in notebook

riedgar-ms committed:
Wrong notebook

riedgar-ms committed:
Didn't mean to modify

riedgar-ms committed:
Fix bullet list references

riedgar-ms committed:
Fix quotation marks

riedgar-ms committed:
That's sort of embarassing....

riedgar-ms committed:
Fixing some things

riedgar-ms committed:
Improve the notebook

riedgar-ms committed:
Typo fix

riedgar-ms committed:
So, we need a return type?

riedgar-ms committed:
Only return the bulleted list

riedgar-ms committed:
Prompt tweaking and updating the notebook.

riedgar-ms committed:
Fiddle, fiddle, fiddle

riedgar-ms committed:
Fiddle, fiddle, fiddle

riedgar-ms committed:
Tweaking prompt

riedgar-ms committed:
Update

riedgar-ms committed:
Fiddling with the planning prompt

riedgar-ms committed:
Prompt engineering

riedgar-ms committed:
Tweak planning metaprompt

riedgar-ms committed:
Rename notebook in view of another which has been added

riedgar-ms committed:
In process demonstration

riedgar-ms committed:
Prompt tweak

riedgar-ms committed:
Update with some sample data

riedgar-ms committed:
Drafting an extra skill

riedgar-ms committed:
Trying to get the new function working

riedgar-ms committed:
Some simple debugging info

riedgar-ms committed:
Simplify a prompt

riedgar-ms committed:
Update notebook

riedgar-ms committed:
Make a new set of skills to pass needed arguments along

riedgar-ms committed:
Push execution sample

riedgar-ms committed:
Show some output

riedgar-ms committed:
Typo fixing

riedgar-ms committed:
Remove GroundingSkill

riedgar-ms committed:
Rename Grounding2Skill

riedgar-ms committed:
Skill tweaks and notebook updates

riedgar-ms committed:
Update notebook

riedgar-ms committed:
Address spelling complaints

lemillermicrosoft committed:
👩‍💻 Update kernel syntax, grounding skill, and add groundedness checks example

This commit updates the kernel syntax examples in the dotnet sample
project to comment out some of the examples that are not needed for
testing. It also modifies the grounding skill to use separate parameters
for the input text, the reference context, and the entities to check.
The skprompt files for the grounding skill are updated accordingly.

Additionally, this commit adds a new example file for demonstrating how
to use the GroundingSkill to remove the entities which are not grounded
in the original text. The example uses the KernelBuilder to import the
GroundingSkill and the SummarizeSkill from the sample skills directory,
and then uses the SequentialPlanner to create and invoke a plan that
takes a text input and returns a grounded summary. The example also
shows how to use the TextSkill to print the results

lemillermicrosoft committed:
🔧 Add entities variable to grounding check

This commit adds a new variable to the context of the reference check
method, which contains the extracted entities from the summary text.
This variable is used by the reference check method to perform
grounding checks on the entities and ensure they are consistent with
the reference context. This improves the accuracy and reliability of
the reference check method and the quality of the summaries.

lemillermicrosoft committed:
🐛 Fix groundedness checks in kernel syntax examples

The previous code for the groundedness checks in the kernel syntax
examples had some errors and inconsistencies. This commit fixes the
following issues:

- The extractionResult and groundingResult variables were not properly
  cast to strings before being used as arguments or printed to the
  console.
- The entity_excision function was invoked with the wrong argument,
  passing the groundingResult instead of the context.
- The final output message was missing the word "Planning" to indicate
  the task.

lemillermicrosoft committed:
Merge branch 'riedgar-ms/grounding-skill' into 613_grounding

lemillermicrosoft committed:
Undo program.cs changes

lemillermicrosoft committed:
📝 Update groundedness checks example

This commit updates the groundedness checks example in the kernel
syntax examples folder. It adds some comments and example output to
the code, and removes an unnecessary line that printed the ask
variable. It also improves the formatting of the ask variable by
removing the extra quotes and using a single line string. The commit
aims to make the example more clear and readable for the users.

riedgar-ms commented on samples/skills/GroundingSkill/ReferenceCheckEntities/skprompt.txt:
```diff
@@ -55,13 +55,18 @@ And a realio, trulio, little pet dragon
 </grounding_context>
 
-Please put your response after the "Response:". Produce a bulleted list of the <ungrounded_entities> which are not grounded in the context, and then repeat the <context>
+Please put your response after the "Response:". Produce a bulleted list of the <ungrounded_entities> which are not grounded in the context.
```
Given this change, does the example need updating above?

riedgar-ms replied:
(Sorry, GitHub won't let me highlight the lines directly.... L46-51

lemillermicrosoft replied:
Good catch!

riedgar-ms commented on samples/skills/GroundingSkill/ReferenceCheckEntities/config.json:
```diff
@@ -13,13 +13,18 @@
         "defaultValue": ""
+      },
+      {
+        "name": "entities",
```
How does this get into the context and made available?

riedgar-ms commented on samples/dotnet/kernel-syntax-examples/Example48_GroundednessChecks.cs:
```diff
@@ -0,0 +1,202 @@
+        var plan = await planner.CreatePlanAsync(ask);
+        Console.WriteLine(plan.ToPlanString());
+
+        var results = await plan.InvokeAsync(s_groundingText);
```
This seems to be rather different from how plans in Python work? The plan there has no 'invoke' method. And the planner itself just takes a plan and a kernel. I don't see how to pass in the grounding text as an argument to the plan like you're doing here.

riedgar-ms reviewed (commented):
Thanks, but I'm having trouble figuring out how to bring this over to Python

lemillermicrosoft committed:
📝 Remove context from skprompt.txt

The context section in skprompt.txt was redundant and irrelevant for the
task of parsing references to items. This commit removes the context
section and updates the task description accordingly. This simplifies the
task and makes it more focused on the reference check entities skill.
